Analyzer and fixer that recommend case insensitive string comparison

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/78606

The analyzer and fixer are both functional in the happy paths and some edge cases, so it's good to start reviewing.

TODOs:
- [x] Handle additional `IndexOf` overloads that take a `StringComparison`.
- [x] Handle cases where the first invocation is parenthesized: https://github.com/dotnet/roslyn-analyzers/pull/6662#discussion_r1213555865
- [x] Add tests where there is no diagnostic expected.
- [x] Add tests where the string is a more complicated expression.
- [x] Handle cases where the `ToLower{Invariant}` and `ToUpper{Invariant}` invocation is inverted. For example, instead of `a.Compare(b.ToLower())` or `a.CompareTo(b.ToUpper())`.
- [x] Handle named parameters where applicable.
- [x] Change title.
- [x] `CompareTo` should just be analyzed but not fixed.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,205 @@
+        internal const string StringStartsWithMethodName = "StartsWith";
+        internal const string StringCompareToMethodName = "CompareTo";
+
+        private static readonly LocalizableString s_localizableStringComparisonTitle = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparisonTitle));
```
I think we inline localizable strings when they are used once only. See https://github.com/dotnet/roslyn-analyzers/pull/5406

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,205 @@
+        {
+            // Retrieve the essential types: string, StringComparison, StringComparer
+
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemString, out INamedTypeSymbol? stringType))
```
I'd do it as `context.Compilation.GetSpecialType(SpecialType.System_String)`. It's never null, but *could* be error symbol.

carlossanlop replied:
> but _could_ be error symbol

Interesting. Do you have any idea why?

carlossanlop replied:
> context.Compilation.GetSpecialType(SpecialType.System_String)

If this is preferred for special types, we could create an analyzer to suggest that 😄

carlossanlop replied:
Suggestion applied.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,205 @@
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation caseChangingInvocation = (IInvocationOperation)context.Operation;
```
```suggestion
                var caseChangingInvocation = (IInvocationOperation)context.Operation;
```

carlossanlop replied:
I prefer to use explicit.

carlossanlop replied:
Which made me realize I accidentally used `var` in a couple of places. For consistently, I'll use explicit in those locations.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,205 @@
+                    return;
+                }
+
+                context.ReportDiagnostic(diagnosableInvocation.CreateDiagnostic(rule, diagnosableMethod.Name));
```
One of the two rules have argument, and the other doesn't. I personally wouldn't pass the argument in both cases.

carlossanlop replied:
Good catch. I fixed this by moving the diagnostic creation to the above if else.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+    /// <summary>
+    /// CA1862: Prefer the StringComparison method overloads to perform case-insensitive string comparisons.
+    /// </summary>
+    public class RecommendCaseInsensitiveStringComparisonFixer : CodeFixProvider
```
Missing ExportCodeFixProvider and Shared attributes.
Consider sealing as well.

carlossanlop replied:
Good catches. I was originally assuming I would need this to be abstract so that I could create a C# and a VB implementation, but I didn't need them in the end (so far). But forgot to add the attribute here.

Since I still have some TODOs left, I wouldn't be surprised if I ended up needing the C# and VB classes in the end. For now, I'll address your suggestion.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            CancellationToken ct = context.CancellationToken;
+            if (ct.IsCancellationRequested)
+            {
+                return;
+            }
```
I have never seen this pattern in the start of RegisterCodeFixesAsync

carlossanlop replied:
You're correct, me neither, but I was wondering about this yesterday. I recently worked on writing async code in dotnet/runtime and @stephentoub taught me some tricks.

@stephentoub what do you think about using the cancellation detection at the beginning of async methods to exit early? Do you know if there could be a reason not to use it in this repo?

stephentoub replied:
Does this even do the right thing?  The expectation of cancelable Task-returning methods is that, if they're canceled, the Task ends in a Canceled state... this is going to look to the caller like it completed successfully even if it completed due to cancellation without actually doing its job.

Let's delete this.  If it's really important to do an up-front check for cancellation, you can add `context.ThrowIfCancellationRequested();` here instead, but if we're not doing that elsewhere in the repo, we shouldn't do so here, either.

carlossanlop replied:
One thing I'm remembering I was told is that the cancellation detection is usually only done inside public methods, which is the case here.

I also used it in line 109 (a private method), maybe I could remove that one.

stephentoub replied:
> One thing I'm remembering I was told is that the cancellation detection is usually only done inside public methods, which is the case here.

I'm not sure who told you what, but this isn't accurate.

carlossanlop replied:
> I'm not sure who told you what, but this isn't accurate. It is not the case that public methods must check the supplied cancellationToken manually, nor is it the case that non-public methods can't.

You're correct, I apologize, I completely misremembered. The explanations you gave me were around exceptions, not cancellation tokens, and I was handling them incorrectly: https://github.com/dotnet/runtime/pull/70574

> Let's delete this. If it's really important to do an up-front check for cancellation, you can add context.ThrowIfCancellationRequested(); here instead, but if we're not doing that elsewhere in the repo, we shouldn't do so here, either.

Ok thanks. Will remove.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+
+            // Only diagnose overloads that take one string argument:
+            // Contains(string), IndexOf(string), StartsWith(string), CompareTo(string)
+            if (invocation.Arguments.Count() != 1)
```
Isn't there a Length property instead of LINQ's Count method?

carlossanlop replied:
I can change it.

I wonder if the perf is noticeably better in Length over Count, so that we could consider suggesting an analyzer+fixer for that.

I thought we already had one, but the closest one I found was [CA1836: "Prefer IsEmpty over Count or Length"](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1836).

Youssef1313 replied:
@carlossanlop That would be https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1829.

carlossanlop replied:
Ok interesting. `invocation.Arguments.Count() != 1` did not give me the suggestion to use `Length`. The analyzer does not seem to diagnose `ImmutableArray`. I changed it anyway, but I wonder if the behavior is expected or not.

@dotnet/area-system-collections Hey folks, do you know the reason why?

carlossanlop replied:
I opened an issue to track this: https://github.com/dotnet/roslyn-analyzers/issues/6667

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs:
```diff
@@ -0,0 +1,488 @@
+            DiagnosticResult diagnosticResult = VerifyCS
+                .Diagnostic(rule)
+                .WithArguments(argument)
+                .WithSpan(startLine, startColumn, endLine, endColumn);
```
This overload should be avoided. Instead, the markup syntax should be used `[| |]` like this:

https://github.com/dotnet/roslyn-analyzers/blob/2b6ab8d727ce73a78bcbf026ac75ea8a7c804daf/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs#L111

carlossanlop replied:
Totally forgot about this pattern. I copied the pattern used in one of my old analyzers. I'll switch to the suggested one.

carlossanlop replied:
Ok I switched, but I am unsure if I can still specify whifch rule to expect. Do you know if it's possible? I haven't been able to find code examples doing this.

Also, I had to use MarkupOptions.UseFirstDescriptor to be able to detect a single rule. Unsure if this is the best practice. Let me know if you know.

Youssef1313 replied:
If you want to assert the argument (as it is done currently), you can go with `{|#0:code|}` syntax instead. Example:

https://github.com/dotnet/roslyn-analyzers/blob/2b6ab8d727ce73a78bcbf026ac75ea8a7c804daf/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureTests.Enums.cs#L78-L111

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+                return;
+            }
+
+            string caseChangingMethodName;
```
caseChangingPropertyName?

carlossanlop replied:
Maybe I should use a different word. I was using "method" here as synonym of "technique" or "approach", not referring to "source code function".

The case changing string stored in this variable is used to retrieve either enum values from `StringComparison` or the properties from `StringComparer`.

I'll change this to `caseChangingApproachName`.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+                CodeAction.Create(
+                    title: MicrosoftNetCoreAnalyzersResources.RecommendCaseInsensitiveStringComparisonTitle,
+                    createChangedDocument,
+                    equivalenceKey: MicrosoftNetCoreAnalyzersResources.RecommendCaseInsensitiveStringComparisonTitle + invocation.TargetMethod.Name + caseChangingMethodName),
```
Wouldn't this require few FixAll passes to actually fix all diagnostics?

carlossanlop replied:
Can you please elaborate? I'm not very familiar with this.

carlossanlop replied:
I see. So I guess I need to change the equivalenceKey string to something that would work better with FixAll, correct? If that's the case, what is the recommended format to use for such string?

Youssef1313 replied:
`nameof(MicrosoftNetCoreAnalyzersResources.RecommendCaseInsensitiveStringComparerStringComparisonCodeFixTitle)` will work well.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    title: MicrosoftNetCoreAnalyzersResources.RecommendCaseInsensitiveStringComparisonTitle,
```
I wouldn't use the analyzer title as the codefix title, and would make the codefix title a bit shorter.

carlossanlop replied:
I welcome suggestions. What would you use?

Youssef1313 replied:
Finding good titles is hard 😄 
Maybe:

- `Use '{0}' with 'StringComparison.{1}'` where `{0}` will be `Contains`, `IndexOf`, or `StartsWith`
- `Use 'StringComparer.{0}.Compare'

carlossanlop replied:
Just to confirm: You're not suggesting that I change the actual resource, you're suggesting that I literally use the string you provided as the value for the title argument. Is this accurate?

carlossanlop replied:
Sorry, nevermind. I thought I was typing on your equivalenceKey suggestion. Ignore.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs:
```diff
@@ -0,0 +1,488 @@
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = StringComparer.InvariantCultureIgnoreCase.Compare(a, b);
```
Consider testing named arguments, also with the order swapped (I don't expect this to be in real code though)

carlossanlop replied:
Yep. It's in my TODO list in the description.

carlossanlop replied:
Addressed! Phew that was... interesting.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs:
```diff
@@ -0,0 +1,488 @@
+        }
+
+        [Fact]
+        public Task Diagnostic_CompareTo_ToUpperInvariant_Assign()
```
Usually, we make test methods `async` and await the task instead of returning it.

carlossanlop replied:
There's no need for that if the only asynchronous item inside your method is the object that is being returned. You can defer the awaiting to whoever is calling your method.

If I was doing more awaits across my method, I would not have a choice but to make my method async.

carlossanlop replied:
I'd like to double check with the expert. @stephentoub would you mind correcting me if what I said was not accurate?

stephentoub replied:
The reason to make the method non-async and return the task would be performance, saving an extra state machine.  That can be worthwhile in production code.  It is not worthwhile here in a test.  In particular, by not having the extra state machine, if the test fails and you get an exception, the exception stack trace wouldn't have this test method in the call stack.

carlossanlop replied:
Thank you for the explanation. I'll await the methods here.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+            }
+
+            string caseChangingMethodName;
+            switch (instanceOperation.TargetMethod.Name)
+            {
+                case RecommendCaseInsensitiveStringComparisonAnalyzer.StringToLowerMethodName or RecommendCaseInsensitiveStringComparisonAnalyzer.StringToUpperMethodName:
+                    caseChangingMethodName = RecommendCaseInsensitiveStringComparisonAnalyzer.StringComparisonCurrentCultureIgnoreCaseName;
+                    break;
+                case RecommendCaseInsensitiveStringComparisonAnalyzer.StringToLowerInvariantMethodName or RecommendCaseInsensitiveStringComparisonAnalyzer.StringToUpperInvariantMethodName:
+                    caseChangingMethodName = RecommendCaseInsensitiveStringComparisonAnalyzer.StringComparisonInvariantCultureIgnoreCaseName;
+                    break;
+                default:
+                    return;
+            }
```
Consider moving this into `FixInvocationAsync`

carlossanlop replied:
Would you mind explaining why? Shouldn't we try to collect all the metadata in this method, and return when it's not possible, before moving to FixInvocationAsync?

Youssef1313 replied:
Yup, I was missing that there is a `default` case that returns, but is this case actually reachable?

carlossanlop replied:
I don't expect the default to be reachable, but I prefer to explicitly mention the two possible cases for clarity, and the default with return is just defensive.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,201 @@
+            Task<Document> createChangedDocument(CancellationToken _) => FixInvocationAsync(doc, root,
+                invocation, instanceOperation,
+                stringComparisonType, stringComparerType,
+                invocation.TargetMethod.Name, caseChangingMethodName, ct);
```
`ct` doesn't feel right. I think we should use the CancellationToken provided to the lambda (which is currently discarded)

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,188 @@
+                    return;
+                }
+
+                if (caseChangingInvocation.Parent is not IInvocationOperation diagnosableInvocation)
```
Quite uncommon case, but should we walk up parenthesized operations first before checking the diagnosable invocation?

carlossanlop replied:
Can you give me an example?

Youssef1313 replied:
@carlossanlop `int r = (a.ToLower()).CompareTo(b);`

carlossanlop replied:
Ah good point! Let me test this and determine how complex it would be to diagnose a case like this.

carlossanlop replied:
Interesting, the code works as it is without any additional changes. I'll push the test so you can see what I mean.

carlossanlop replied:
Which means I was inadvertently removing parenthesis. I am trying to think of a case where the parenthesis _need_ to be preserved.

carlossanlop replied:
Ah, it only works for C#, but it doesn't work for VB.

carlossanlop replied:
Fixed!

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs:
```diff
@@ -0,0 +1,488 @@
+        [Fact]
+        public Task Diagnostic_CompareTo_ToLower_Assign()
+        {
+            string originalCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = a.ToLower().CompareTo(b);
+        return r;
+    }
+}";
+            string fixedCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = StringComparer.CurrentCultureIgnoreCase.Compare(a, b);
+        return r;
+    }
+}";
```
I think fixed code changes semantics.

```csharp
using System;

string a = "a";
string b = "A";

Console.WriteLine(a.ToLower().CompareTo(b)); // Compares "a" to "A", gives -1
Console.WriteLine(StringComparer.CurrentCultureIgnoreCase.Compare(a, b)); // Ignores case, so gives 0
```

carlossanlop replied:
I asked the question in the main issue https://github.com/dotnet/runtime/issues/78606#issuecomment-1572625053

carlossanlop replied:
Ok, so for now, I'll only diagnose `CompareTo`, but won't offer a fix. We can address this later if we want to.

carlossanlop replied:
@Youssef1313 what's the best way to determine if an analyzer triggered, but nothing got fixed? Like this?:

```cs
        private Task YesDiagnosticButNoFix(string originalSource)
        {
            VerifyVB.Test test = new()
            {
                TestCode = originalSource,
                FixedCode = originalsource
            };
            return test.RunAsync();
        }
```

The reason I ask is because I am looking at other examples, and it seems that when we don't expect a _diagnostic_ at all, we just pass a `TestCode` string, but no `FixedCode`, like this:

```cs
        private Task NoDiagnostic(string originalSource)
        {
            VerifyVB.Test test = new()
            {
                TestCode = originalSource,
            };
            return test.RunAsync();
        }
```

Youssef1313 replied:
When there is no diagnostic, both snippets should be equivalent.
When there is a diagnostic, the first should be used as it asserts that no code fix is offered

carlossanlop replied:
Ah okay, my examples were inverted. Thanks for the confirmation.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -199,7 +199,14 @@ private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
                 }
 
-                if (caseChangingInvocation.Parent is not IInvocationOperation diagnosableInvocation)
+                // Ignore parenthesized operations
+                IOperation? ancestor = caseChangingInvocation.Parent;
+                while (ancestor is not null and IParenthesizedOperation)
+                {
+                    ancestor = ancestor.Parent;
+                }
```
There is already an `WalkUpParentheses` extension method.

carlossanlop replied:
Good to know. Let me use it.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -76,10 +76,17 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
             }
 
+            // Ignore parenthesized operations
+            IOperation? instanceGenericOperation = invocation.Instance;
```
I think the codefix does too much. The analyzer can add a properties dictionary to the diagnostic which can be retrieved here.

carlossanlop replied:
Do you have an example? I don't know how to do what you're suggesting.

carlossanlop replied:
I _think_ you're referring to the fact that I can tell the analyzer where to put squiggles , but I can also indicate what needs to be changed by the fixer, so that the fixer doesn't need to do the whole search again. Correct?

Youssef1313 replied:
Example, in the analyzer:

https://github.com/dotnet/roslyn-analyzers/blob/2b6ab8d727ce73a78bcbf026ac75ea8a7c804daf/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs#L139-L142

In the codefix:

https://github.com/dotnet/roslyn-analyzers/blob/2b6ab8d727ce73a78bcbf026ac75ea8a7c804daf/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.Fixer.cs#L42

Basically you just add "more info" to the diagnostic as a `Dictionary<string, string>` which is retrievable in the codefix. This avoids the need to re-calculate things that were already calculated in the analyzer.

carlossanlop replied:
Thanks.

Things I think I can remove from the fixer because it might not be necessary to verify them:

- These checks that are probably unnecessary: https://github.com/dotnet/roslyn-analyzers/blob/59c0115ec26fb64e0db8f2883a267dbe312bbee6/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs#L58-L73
- This debug assert: https://github.com/dotnet/roslyn-analyzers/blob/59c0115ec26fb64e0db8f2883a267dbe312bbee6/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs#L107

Things I can't remove:

- This isn't a string: https://github.com/dotnet/roslyn-analyzers/blob/59c0115ec26fb64e0db8f2883a267dbe312bbee6/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs#L50-L54
- The parenthesis operand digging: https://github.com/dotnet/roslyn-analyzers/blob/59c0115ec26fb64e0db8f2883a267dbe312bbee6/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs#L76-L80
- The instanceOperation and the caseChangingMethodName, they are both passed down to the fixer method and the second one isn't even calculated in the analyzer: https://github.com/dotnet/roslyn-analyzers/blob/59c0115ec26fb64e0db8f2883a267dbe312bbee6/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs#L85-L101

carlossanlop replied:
And I couldn't find anything to send as properties because it's not useful as a string.

carlossanlop replied:
The one you mentioned at the top is an instance object. Were you suggesting that I pass it somehow in the properties dictionary?

Youssef1313 replied:
Yeah properties can only be strings.
I missed that you need the actual operation, I was thinking only instanceOperation.TargetMethod.Name is relevant. You can keep as it's now.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs:
```diff
@@ -0,0 +1,229 @@
+            {
+                TestCode = originalSource,
+                FixedCode = fixedSource,
+                MarkupOptions = MarkupOptions.UseFirstDescriptor
```
This shouldn't be needed.

carlossanlop replied:
The exception message says:

```
Multiple diagnostic descriptors with ID {diagnosticId} were found
```

I interpreted it differently: you have more than one descriptor, they all share the same ID. Which is my case.

Youssef1313 replied:
Indeed. I read it completely incorrectly.

carlossanlop commented:
The CI keeps failing saying that I should not have modified Microsoft.CodeAnalysis.NetAnalyzers.md and .sarif. But if I revert the changes I manually added, and build, I keep getting the same error. So I'm not sure how to edit those files (or get them automatically generated).

Youssef1313 commented:
> The CI keeps failing saying that I should not have modified Microsoft.CodeAnalysis.NetAnalyzers.md and .sarif. But if I revert the changes I manually added, and build, I keep getting the same error. So I'm not sure how to edit those files (or get them automatically generated).

Did you run `dotnet msbuild -t:pack`?

carlossanlop commented:
> Did you run `dotnet msbuild -t:pack`?

Yes, but it didn't work at first. Now I did a `git clean -fdx` and ran everything from scratch, and that worked. Thanks.

carlossanlop commented:
I'm currently working on named arguments. I forgot how complex it can be 😄 .

carlossanlop commented:
Addressed all the TODOs. I am now going to run this against dotnet/runtime.

carlossanlop commented:
Nice, I didn't find any instances that need to be fixed in dotnet/runtime. To be sure it wasn't a problem of not loading the analyzer assembly correctly, I double checked by adding one diagnosable invocation myself, and it does trigger a build error as expected. Otherwise, the build succeeds.

I'll now try it in other repos.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,204 @@
+    /// <summary>
+    /// CA1862: Prefer the StringComparison method overloads to perform case-insensitive string comparisons.
+    /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = RCISCAnalyzer.RuleId), Shared]
```
I think this needs to move to derived types.

mavasani replied:
+1

carlossanlop replied:
I assume `[Shared]` stays here?

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpRecommendCaseInsensitiveStringComparisonFixer.cs:
```diff
@@ -0,0 +1,98 @@
+{
+    using RCISCAnalyzer = RecommendCaseInsensitiveStringComparisonAnalyzer;
+
+    internal class CSharpRecommendCaseInsensitiveStringComparisonFixer : RecommendCaseInsensitiveStringComparisonFixer
```
```suggestion
    internal sealed class CSharpRecommendCaseInsensitiveStringComparisonFixer : RecommendCaseInsensitiveStringComparisonFixer
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,204 @@
+            INamedTypeSymbol stringComparisonType, out SyntaxNode? mainInvocationInstance);
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(RCISCAnalyzer.RuleId);
+        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
```
```suggestion

        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -2054,4 +2054,25 @@ Widening and user defined conversions are not supported with generic types.</val
     <value>Prefer an 'IsEmpty' check rather than using 'Any()', both for clarity and for performance</value>
   </data>
+  <data name="RecommendCaseInsensitiveStringComparerStringComparisonCodeFixTitle" xml:space="preserve">
+    <value>Use the string.{0}(string, StringComparison) overload</value>
```
```suggestion
    <value>Use the 'string.{0}(string, StringComparison)' overload</value>
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -2054,4 +2054,25 @@ Widening and user defined conversions are not supported with generic types.</val
+    <value>Prefer using 'StringComparer' to perform case-insensitive string comparisons</value>
+  </data>
+  <data name="RecommendCaseInsensitiveStringComparerDescription" xml:space="preserve">
+    <value>Avoid calling 'ToLower', 'ToUpper', 'ToLowerInvariant' and 'ToUpperInvariant' to perform case-insensitive string comparisons when using 'CompareTo'. Instead, use 'StringComparer' to perform case-insensitive comparisons.</value>
```
Might be good to also add that the ToLower/ToUpper methods lead to an allocation, which is saved by switching to the recommended comparison method.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,238 @@
+                {
+                    diagnosableInvocation = argumentInvocationAncestor;
+                }
+                else
```
Should we consider detecting simple patterns across statements, such as below?

```csharp
var loweredStr = str.ToLower();
if (loweredStr.Contains(anotherStr))
```

We would need to check that we have exactly one reference to the temp variable. If you are interested, we already added another analyzer which implements a simple heuristic for above patterns: https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfAnalyzer.cs. If not, you may also feel free to also file a tracking issue for this enhancement in future if you deem it to be worthwhile.

carlossanlop replied:
Yes, V2. Opened: https://github.com/dotnet/roslyn-analyzers/issues/6689

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs:
```diff
@@ -0,0 +1,204 @@
+            //if (mainInstanceOperation is IInvocationOperation instanceInvocation &&
+            //    instanceInvocation.TargetMethod.Name is
+            //        RCISCAnalyzer.StringToLowerMethodName or RCISCAnalyzer.StringToUpperMethodName or
+            //        RCISCAnalyzer.StringToLowerInvariantMethodName or RCISCAnalyzer.StringToUpperInvariantMethodName)
+            //{
+            //    removableInvocation = instanceInvocation;
+            //}
+            //// The other option is that the offending operation is the first argument
+            //// a.Contains(b.ToLower(), ...)
+            //else
+            //{
+            //    IOperation? argumentOffendingOperation = mainInvocation.Arguments.FirstOrDefault();
+            //    while (argumentOffendingOperation is not null and IParenthesizedOperation parenthesizedOperation)
+            //    {
+            //        argumentOffendingOperation = parenthesizedOperation.Operand;
+            //    }
+
+            //    // TODO: Do not use firstordefault, add test that puts "value" in another position
+            //    if (argumentOffendingOperation is IArgumentOperation argumentOperation &&
+            //        argumentOperation.Children.FirstOrDefault() is IInvocationOperation argumentInvocation)
+            //    {
+            //        removableInvocation = argumentInvocation;
+            //        isChangingCaseInArgument = true;
+            //    }
+            //}
+
+            //string caseChangingApproachName;
+            //if (removableInvocation.TargetMethod.Name is
+            //    RCISCAnalyzer.StringToLowerMethodName or RCISCAnalyzer.StringToUpperMethodName)
+            //{
+            //    caseChangingApproachName = RCISCAnalyzer.StringComparisonCurrentCultureIgnoreCaseName;
+            //}
+            //else
+            //{
+            //    Debug.Assert(removableInvocation.TargetMethod.Name is
+            //        RCISCAnalyzer.StringToLowerInvariantMethodName or RCISCAnalyzer.StringToUpperInvariantMethodName);
+
+            //    caseChangingApproachName = RCISCAnalyzer.StringComparisonInvariantCultureIgnoreCaseName;
+            //}
+            */
```
Remove?

mavasani reviewed (commented):
Overall, looks great. I have reviewed the analyzer and the tests properly, and skimmed through the fixer code. Thank you for the exhaustive unit tests!

carlossanlop commented:
I addressed your latest feedback, @Youssef1313 and @mavasani.

There's one CI failure (not sure if related to my change, I'm investigating), but putting that aside, does this look good enough for a sign-off?

Note that I'm going to start working on this one next: https://github.com/dotnet/runtime/issues/78607 . It's closely related to this one. In the API review it was requested that, if possible, we join both this analyzer and that one under the same ID.

mavasani reviewed (approved):
LGTM, thanks!

@Youssef1313 - would you like to do another review pass? If you are able to do so this week, we can wait for your signoff before merging. However, if you are busy, we should go ahead with the merge and you can review whenever convenient. Please let us know, thanks!

mavasani commented:
> There's one CI failure 

@carlossanlop You'd need to re-run `msbuild /t:pack` at repo root and push the changes.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs:
```diff
@@ -0,0 +1,238 @@
+                {
+                    context.ReportDiagnostic(diagnosableInvocation.CreateDiagnostic(RecommendCaseInsensitiveStringComparisonRule, diagnosableMethod.Name));
+                }
+                else if (diagnosableMethod.Equals(compareToStringMethod))
+                {
+                    context.ReportDiagnostic(diagnosableInvocation.CreateDiagnostic(RecommendCaseInsensitiveStringComparerRule));
+                }
```
Not sure if this will be a false-negative in some cases. Previously, I mentioned that the codefix produces code that is semantically different. Is it a false negative in the first place?

Or maybe report CompareTo under a separate ID to give user more control to disable it alone?

carlossanlop replied:
I addressed your concern about CompareTo changing the semantics. I special-cased CompareTo so that it only triggers the analyzer but not the codefix. Check out the unit tests to see how I separate them.

Youssef1313 replied:
Yup, this comment is about whether we should actually warn or not, and if we should warn, whether it should be a separate ID or not.
Per https://github.com/dotnet/roslyn-analyzers/pull/6662#issuecomment-1597519573, I'm personally okay with doing this in a follow-up (unless @mavasani thinks this should be in this PR)

carlossanlop replied:
Ah got it. Yes, let's continue the discussion separately.

Youssef1313 reviewed (approved):
LGTM, with a small comment on CompareTo

Youssef1313 reviewed (commented):
For `CompareTo`, I think we should only report when the case changing method is called on both strings (or a constant known to be lower-case at compile-time). In this case, I *think* the codefix will produce the correct result and we can add it back.

carlossanlop commented:
> For `CompareTo`, I think we should only report when the case changing method is called on both strings (or a constant known to be lower-case at compile-time). In this case, I _think_ the codefix will produce the correct result and we can add it back.

Good point. If you don't have any objections, allow me to do this in V2. Right now, CompareTo is being diagnosed only, no code fix is being offered.

Youssef1313 commented:
> > For `CompareTo`, I think we should only report when the case changing method is called on both strings (or a constant known to be lower-case at compile-time). In this case, I _think_ the codefix will produce the correct result and we can add it back.
> 
> Good point. If you don't have any objections, allow me to do this in V2. Right now, CompareTo is being diagnosed only, no code fix is being offered.

No problem for me to do it in a follow-up. @mavasani thoughts?

carlossanlop commented:
> No problem for me to do it in a follow-up. @mavasani thoughts?

I consulted with @mavasani via chat and he agrees. @Youssef1313, I added your latest CompareTo suggestion to my follow up issue: https://github.com/dotnet/roslyn-analyzers/issues/6689 

I addressed the `msbuild /t:pack` issue and the build should pass now. For future reference for myself, here's what I had to do:

- `git clean -fdx`
- Delete all the generated xlfs from `src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers.xlf`
- Remove the entry that was previously automatically added in `src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.sarif`
- Run `dotnet restore` from root
- Run `msbuild /t:pack` from root
- Build everything as usual
- Push the added changes
